library/cwm/src/lib/cwm/widget.rb

```diff
@@ -0,0 +1,450 @@
+    include Yast::UIShortcuts
+    include Yast::I18n
+
+    # @return [String] id used for widget
+    attr_accessor :widget_id
+
+    # defines widget type for CWM usage
+    def self.widget_type=(val)
+      define_method(:widget_type) { val }
+    end
+
+    # generates description for CWM with widget. Description is auto filled from defined methods.
+    #
+    # methods used to generate description:
+    #
+    # - `#help` [String] to get translated help text for widget
+    # - `#label` [String] to get translated label text for widget
+    # - `#opt` [Array<Symbol>] to get options passed to widget like `[:hstretch, :vstretch]`
+    # - `#validate` [Boolean (String, Hash)] validate widget value, arguments are
+    #   widget key and event map causing validation, for other validation types, overwrite description
+    # - `#init` [nil (String)] initialize widget. like its value. Parameter is widget id
+    # - `#handle` [Symbol,nil (String, Hash)] handle widget changed value or press.
+    #   return value is usually nil, returning symbol can be used to send different event
+    # - `#store` [nil (String, Hash)] store widget value after user confirm dialog
+    # - `#cleanup` [nil (String)] cleanup after widget is destroyed
+    # @raise [RuntimeError] if required method is not implemented or widget id not set.
+    def description
+      raise "Widget '#{self.class}' does set its widget ID" if widget_id.nil?
+      if !respond_to?(:widget_type)
+        raise "Widget '#{self.class}' does set its widget type"
+      end
+
+      res = {}
+
+      if respond_to?(:help)
+        res["help"] = help
+      else
+        res["no_help"] = ""
+      end
+      res["label"] = label if respond_to?(:label)
+      res["opt"] = opt if respond_to?(:opt)
+      if respond_to?(:validate)
+        res["validate_function"] = validate_method
+        res["validate_type"] = :function
+      end
+      res["init"] = init_method if respond_to?(:init)
+      res["handle"] = handle_method if respond_to?(:handle)
+      res["store"] = store_method if respond_to?(:store)
+      res["cleanup"] = cleanup_method if respond_to?(:cleanup)
+      res["widget"] = widget_type
+
+      res
+    end
+
+    # gets if widget is open for modification
+    def enabled?
+      Yast::UI.QueryWidget(Id(widget_id), :Enabled)
+    end
+
+    # Opens widget for modification
+    def enable
+      Yast::UI.ChangeWidget(Id(widget_id), :Enabled, true)
+    end
+
+    # Closes widget for modification
+    def disable
+      Yast::UI.ChangeWidget(Id(widget_id), :Enabled, false)
+    end
+
+  protected
+
+    # helper to check if event is invoked by this widget
+    def my_event?(widget, event)
+      return widget == event["ID"]
+    end
+
+    # shortcut from Yast namespace to avoid including whole namespace
+    # TODO: kill converts in CWM module, to avoid this workaround for funrefs
+    def fun_ref(*args)
+      Yast::FunRef.new(*args)
+    end
+
+  private
+
+    def init_method
```

Comment:
If it is named `*_method` then I would expect it to be a ::Method. The helpers should be renamed to `*_fun_ref`. All the more important because the `::Method`s are also involved.